scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeExtractors.scala

```diff
@@ -37,11 +39,18 @@ object `:parent:` {
   }
 }
 
-object SelfAnnotation {
+object WithChain {
   def unapply(t: Type.With): Option[Type.With] =
     TreeOps.topTypeWith(t) match {
+      // self types
       case (top: Type.With) `:parent:` (_: Term.Param) `:parent:` (_: Template) =>
         Some(top)
+      // type definitions
+      case (top: Type.With) `:parent:` (_: Defn.Type) =>
```

Comment:
I suspect this pattern can be extended to any Defn or Decl, that will cover val/var/def regardless if they're abstract or not.

Reply:
I'm not sure when we don't want this to match for Type.With, there are other tree nodes like Term.ApplyType where this is equally relevant.

Reply:
Yeah I agree for the most part. 

If I open it up to match all `Type.With`s then one test fails (with good reason IMO):

```scala
// Before:
{ def getAnnotationOwner(
       annotationOwnerLookUp: ScLiteral => Option[
         PsiAnnotationOwner with PsiElement]): Option[PsiAnnotationOwner] = 1 }

// After -- breaks in generic type:
{
  def getAnnotationOwner(
      annotationOwnerLookUp: ScLiteral => Option[PsiAnnotationOwner
        with PsiElement]): Option[PsiAnnotationOwner] = 1
}

// Expected:
{
  def getAnnotationOwner(
      annotationOwnerLookUp: ScLiteral => Option[
          PsiAnnotationOwner with PsiElement]): Option[PsiAnnotationOwner] = 1
}
```

From https://github.com/scalameta/scalafmt/blob/master/scalafmt-tests/src/test/resources/default/TypeWith.stat

Reply:
But even in that scenario, if the type were long enough we probably want a newline instead of violating the column limit. Maybe we can make this split a low priority or high cost etc?